File: Aula9/exercicio4.cpp
#include <iostream> // Biblioteca para entrada e saída de dados
#include <list>     // Biblioteca para listas duplamente ligadas
#include <string>   // Biblioteca para manipulação de strings
using namespace std;

int main() {
    int X;
    cout << "Insira o numero de cidades ";
    cin >> X; // Lê a quantidade de cidades que o usuário vai informar

    // Cria listas para armazenar nomes de cidades e suas populações
    list<string> nomes;
    list<int> habitantes;

    // Entrada de dados: para cada cidade, lê nome e população
    for (int i = 0; i < X; i++) {
        string nome;
        int hab;
        cout << "Nome da cidade " << i+1 << ": ";
        cin >> nome;
        cout << "Quantidade de habitantes: ";
        cin >> hab;
        nomes.push_back(nome);       // Adiciona o nome à lista de nomes
        habitantes.push_back(hab);   // Adiciona a população à lista de habitantes
    }

    // Calcular a população média
    double soma = 0;
    for (int h : habitantes) {  // Percorre todos os habitantes
        soma += h;               // Soma os valores
    }
    double media = soma / X;     // Calcula a média
    cout << "Populacao media: " << media << endl;

    // Listar cidades com população acima da média
    cout << "Cidades acima da media: ";
    list<string>::iterator itNome = nomes.begin();      // Iterador para nomes
    list<int>::iterator ithab = habitantes.begin();    // Iterador para habitantes
    while (itNome != nomes.end() && ithab != habitantes.end()) {
        if (*ithab > media) { // Se a população for maior que a média
            cout << *itNome << " (" << *ithab << " habitantes) ";
        }
        ++itNome;  // Avança o iterador de nomes
        ++ithab;   // Avança o iterador de habitantes
    }

    // Encontrar cidade mais e menos populosa
    itNome = nomes.begin();
    ithab = habitantes.begin();
    string cidadeMais = *itNome;  // Inicializa com a primeira cidade
    string cidadeMenos = *itNome;
    int habMais = *ithab;
    int habMenos = *ithab;

    ++itNome;   // Avança para a próxima cidade
    ++ithab;

    while (itNome != nomes.end() && ithab != habitantes.end()) {
        if (*ithab > habMais) { // Atualiza a cidade mais populosa
            habMais = *ithab;
            cidadeMais = *itNome;
        }
        if (*ithab < habMenos) { // Atualiza a cidade menos populosa
            habMenos = *ithab;
            cidadeMenos = *itNome;
        }
        ++itNome;
        ++ithab;
    }

    cout << "Cidade mais populosa: " << cidadeMais << " (" << habMais << " habitantes)";
    cout << "Cidade menos populosa: " << cidadeMenos << " (" << habMenos << " habitantes)";

    // Remover cidades com população igual a Y
    int Y;
    cout << "Informe uma quantidade de habitantes que deseja remover : ";
    cin >> Y;

    itNome = nomes.begin();
    ithab = habitantes.begin();
    while (itNome != nomes.end() && ithab != habitantes.end()) {
        if (*ithab == Y) { // Se a população for igual a Y
            itNome = nomes.erase(itNome);       // Remove a cidade
            ithab = habitantes.erase(ithab);    // Remove a população correspondente
        } else {
            ++itNome;
            ++ithab;
        }
    }

    // Imprimir lista atualizada de cidades e suas populações
    cout << "\nDicionário atualizado:\n";
    itNome = nomes.begin();
    ithab = habitantes.begin();
    while (itNome != nomes.end() && ithab != habitantes.end()) {
        cout << *itNome << " : " << *ithab << endl;
        ++itNome;
        ++ithab;
    }

    return 0; // Encerra o programa
}

File: Aula9/Exercicio4.cs

using System;
using System.Collections.Generic; // Necessário para usar Dictionary e List

class Program
{
    static void Main()
    {
        // Solicita ao usuário o número de pessoas que deseja cadastrar
        Console.Write("Quantas pessoas deseja inserir? ");
        int X = int.Parse(Console.ReadLine());

        // Cria um dicionário para armazenar nome (chave) e idade (valor)
        Dictionary<string, int> pessoas = new Dictionary<string, int>();

        // Entrada de dados: lê nome e idade de cada pessoa
        for (int i = 0; i < X; i++)
        {
            Console.Write("Nome da pessoa " + (i + 1) + ": ");
            string nome = Console.ReadLine();

            Console.Write("Idade de " + nome + ": ");
            int idade = int.Parse(Console.ReadLine());

            pessoas.Add(nome, idade); // Adiciona ao dicionário
        }

        // Exibe todas as pessoas cadastradas
        Console.WriteLine("\n--- Pessoas cadastradas ---");
        foreach (var p in pessoas)
        {
            Console.WriteLine(p.Key + " -> " + p.Value);
        }

        // 1) Calcular a média das idades
        int soma = 0;
        foreach (var idade in pessoas.Values)
        {
            soma += idade; // Soma todas as idades
        }
        double media = (double)soma / pessoas.Count; // Calcula a média

        Console.WriteLine("\nIdade média: " + media);

        // Exibe pessoas com idade acima da média
        Console.WriteLine("Pessoas com idade acima da média:");
        foreach (var p in pessoas)
        {
            if (p.Value > media)
            {
                Console.WriteLine(p.Key + " (" + p.Value + ")");
            }
        }

        // 2) Encontrar a pessoa mais velha e a mais nova
        string maisVelho = null, maisNovo = null;
        int idadeMax = int.MinValue, idadeMin = int.MaxValue;

        foreach (var p in pessoas)
        {
            if (p.Value > idadeMax) // Atualiza a pessoa mais velha
            {
                idadeMax = p.Value;
                maisVelho = p.Key;
            }
            if (p.Value < idadeMin) // Atualiza a pessoa mais nova
            {
                idadeMin = p.Value;
                maisNovo = p.Key;
            }
        }

        Console.WriteLine("\nPessoa mais velha: " + maisVelho + " (" + idadeMax + ")");
        Console.WriteLine("Pessoa mais nova: " + maisNovo + " (" + idadeMin + ")");

        // 3) Remover pessoas com idade igual a Y
        Console.Write("\nDigite um valor Y de idade para remover: ");
        int Y = int.Parse(Console.ReadLine());

        // Criamos uma lista auxiliar porque não podemos remover itens de um dicionário enquanto iteramos sobre ele
        List<string> remover = new List<string>();

        foreach (var p in pessoas)
        {
            if (p.Value == Y)
            {
                remover.Add(p.Key); // Adiciona à lista de remoção
            }
        }

        // Remove as pessoas listadas da lista auxiliar
        foreach (var nome in remover)
        {
            pessoas.Remove(nome);
        }

        // Exibe o dicionário atualizado
        Console.WriteLine("\n--- Dicionário atualizado ---");
        foreach (var p in pessoas)
        {
            Console.WriteLine(p.Key + " -> " + p.Value);
        }
    }
}
//saudades dos comentarios que tinhas nas aulas 4 e 5.